update current_link without downtime

RobberPhex opened the pull request:
if `current_path` is a link, `deploy_helper` will unlink it and re-link it. That will cause downtime, such as http 404 error.

`os.rename` can avoid this downtime.

gregdek commented:
Thanks @RobberPhex. @ramondelafuente please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit' or 'needs_revision' as appropriate.

ramondelafuente commented on web_infrastructure/deploy_helper.py:
```diff
@@ -337,6 +342,16 @@ def remove_unfinished_builds(self, releases_path):
 
         return changes
 
+    def remove_unfinished_link(self, path):
```
Why have a procedure to remove the temporary link? Won't the temp link be removed during rename?

If you have a good reason for it to exist, I do think this could also support the _module.check_mode_
(i.e. not change anything, but know _if_ something will change)

ramondelafuente commented:
Besides the comment in the source - has this really been causing problems in production? Back-to-back delete/create statements? Just curious :-)

robberphex commented:
@ramondelafuente in origin code, when current link need update(such as fire a new version), we need unlink it and re-link it.
After unlink, and before re-link, the site is unavailable(because current link doesn't exist).

In my new solution, I will create temp link for new version, and rename temp link to current link. This can avoid the unavailable time.

btw, I met this situation when I fire a new version(written by php), so I create this PR.

As for check_mode, sorry for forget it, I will commit it soon.

robberphex commented:
@ramondelafuente I fixed check mode in 8361e2a7db67db503acc3854256afb29d8e26dec

---

But, I noticed that when need create_link, it will mark changed to True, even if current link is new version. so I add commit to fix it(a34312e3f3efc566556d7e038e88001f5ed468ae):
When current link point to new release, it will mark changed to False.

Should I split it to new PR?

ramondelafuente commented:
I was thinking about this a little bit - but having that tmp symlink floating around seems to bring a bit more complexity than needed. How about we do a forced symlink (`ln -s -f`)?

My python-FU is not strong but if os.symlink does not support this we could just run a system command, right?

ramondelafuente commented:
As for the changes regarding check_mode: nice job. That looks good to me now.

robberphex commented:
_I am afraid ln -f is not a good idea_ @ramondelafuente 

Let's read the code of `ln`(http://git.savannah.gnu.org/cgit/coreutils.git/tree/src/ln.c#n329). ln also unlink and re-link, it also cause downtime.

So, I look for rename, as [POSIX](http://pubs.opengroup.org/onlinepubs/009695399/functions/rename.html) described, rename will finish link update in one function.
Next, I find the `os.rename`'s implement([2.6](https://github.com/python/cpython/blob/2.6/Modules/posixmodule.c#L2626), [2.7](https://github.com/python/cpython/blob/2.7/Modules/posixmodule.c#L2808), [3.4](https://github.com/python/cpython/blob/1fe0fd9feb6a4472a9a1b186502eb9c0b2366326/Modules/posixmodule.c#L4146)):
- in 2.x, os.rename will call glibc's rename.
- in 3.x, os.rename will call renameat or rename.

`renameat` is a [syscall](http://sourceware.org/git/?p=glibc.git;a=blob;f=sysdeps/unix/sysv/linux/syscalls.list;hb=78bd7499af46d739ce94410eaeea006e874ca9e5#l92), there is no downtime.
in some implement, `rename` also call [`renameat`](http://sourceware.org/git/?p=glibc.git;a=blob;f=sysdeps/unix/sysv/linux/generic/rename.c;h=ef0e1683921bd8bc9cd5e4556055d4a65f9fe03b;hb=78bd7499af46d739ce94410eaeea006e874ca9e5#l26), there is no downtime.
But some implement also use [_unlink and re-link_](http://sourceware.org/git/?p=glibc.git;a=blob;f=sysdeps/posix/rename.c;h=0f13abbf5d249eabc4c527bbd5ef94185007255e;hb=78bd7499af46d739ce94410eaeea006e874ca9e5#l24) to rename, in low level, I think this is minimal downtime.

robberphex commented:
Just re-commit for correct my author info(avoid using company's email)

ramondelafuente commented:
Seems like you did a lot more homework on this than I did... and it seems like os.rename works better indeed.

Final thought: in order to keep everything contained in the build directory, how about we create the tmplink there, and rename it to the current-path? This way you would also not care if a tmplink remained. (although I'm not sure how unhappy some tools would be with a symlink recursively pointing to the directory it's in).

ramondelafuente commented:
And when it comes to this micro-second downtime during un-link and link, at that level I think moving away from symlinks altogether might be a good idea. Like writing the new path to the webserver config and reloading...

robberphex commented:
I think recursively link is very unfriendly to some tools, and maybe cause additional complexity in other ansible module. So I think crate temp link in project path is better. :)

ramondelafuente commented:
In that case @gregdek shipit!  :tada:

gregdek commented:
Thanks again to @RobberPhex for this PR, and thanks @ramondelafuente for reviewing. Marking for inclusion.

bcoca commented:
@RobberPhex could you flatten the commits?

robberphex commented:
I rebase it just now.

@bcoca my three commit in one line( no merge), what's means "flatten"? Did you mean squash three commit to one?

bcoca commented:
merge the commits down to a single one, to make cherry picking easy

robberphex commented:
@bcoca done.
